### PR TITLE
two fixups for the arrayop implementation

### PR DESCRIPTION
**Pull request by MartinNowak**
- add support for mixed vector types
- fix argument index in initScalarVecs

### CODE REVIEW

File: src/core/internal/arrayop.d [hunk 1]
```diff
@@ -224,7 +283,7 @@ string scalarExp(Args...)()
         }
         else static if (isBinaryOp(arg))
         {
-            stack[$ - 2] = "(cast(T)(" ~ stack[$ - 2] ~ " " ~ arg ~ " " ~ stack[$ - 1] ~ "))";
```
Comment (MartinNowak) [inline src/core/internal/arrayop.d]: The existing arrayop behavior only casts the whole RHS before assignment, but not intermediate expressions. This is fixed here.

File: src/core/internal/arrayop.d [hunk 2]
```diff
@@ -302,6 +362,25 @@ template not(alias tmlp)
 {
     enum not(Args...) = !tmlp!Args;
 }
+/// find index for which pred!haystack[index] is true or return -1
```
Comment (WalterBright) [inline src/core/internal/arrayop.d]: Please use Ddoc style comments, with `Params:` and `Returns:` sections. It's not just for readability, it's so tools can be used on the documentation.

File: src/core/internal/traits.d [hunk 3]
```diff
@@ -210,3 +210,23 @@ template Filter(alias pred, TList...)
                 Filter!(pred, TList[$/2 ..  $ ]));
     }
 }
+
+// std.meta.staticMap
```
Comment (WalterBright) [inline src/core/internal/traits.d]: Looking at the documentation for std.meta.staticMap, all it says is:
```
Evaluates to $(D AliasSeq!(F!(T[0]), F!(T[1]), ..., F!(T[$ - 1]))).
```
but that is not what this template body does.

File: src/core/internal/arrayop.d [hunk 4]
```diff
@@ -224,7 +283,7 @@ string scalarExp(Args...)()
         }
         else static if (isBinaryOp(arg))
         {
-            stack[$ - 2] = "(cast(T)(" ~ stack[$ - 2] ~ " " ~ arg ~ " " ~ stack[$ - 1] ~ "))";
+            stack[$ - 2] = "(" ~ stack[$ - 2] ~ " " ~ arg ~ " " ~ stack[$ - 1] ~ ")";
```
Comment (WalterBright) [inline src/core/internal/arrayop.d]: Any way this and the following red lines can be turned to green?

Comment (MartinNowak) [inline src/core/internal/arrayop.d]: No, it's CTFE only functions.

File: src/core/internal/traits.d (hunk 3 above)
Comment (MartinNowak) [inline src/core/internal/traits.d]: It's exactly what this template does, in fact I've copied the implementation.
We still need to rename TypeTuple to AliasSeq in this module though.

File: src/core/internal/arrayop.d (hunk 2 above)
Comment (MartinNowak) [inline src/core/internal/arrayop.d]: Done.
Note that this already was a ddoc comment and also it's just a private internal helper template with only 5 lines of implementation.

**Comment by MartinNowak** [comment]
Done, thanks for the review @WalterBright.

**Pull request merged**